docs/proposals/generic-schema-batching-proposal.md

```diff
@@ -0,0 +1,987 @@
+
+4. **Brittle Architecture**: Adding a new schema type requires changes across multiple files:
+   - Schema constants
+   - DTOs
+   - Validators
+   - Announcement processors
+   - API controllers
+   - Queue mappings
+
+5. **REST API ≠ On-Chain Format**: The current API accepts DSNP Activity Content format but must manually transform it to Parquet format, limiting generic schema support.
+
+### Requirements
+
+**Must Have**:
+- Support batching of any Parquet-based schema
+- Backward compatibility with existing DSNP endpoints
+- Dynamic schema discovery from blockchain
+- Support for non-DSNP schemas (AT Protocol, custom)
+
+**Should Have**:
+- Generic validation framework (different issue/PR?)
+- Gradual migration path (and/or backwards compatibility)
+- Plugin architecture for complex (non-DSNP)schemas
+- Comprehensive API documentation on how to use this feature
+
+---
+
+## Current Architecture Analysis
+
+### Batching Flow (Current State)
+
+```
+User → REST API Endpoint (/broadcast, /reply, etc.)
+  ↓
+DTO Validation (BroadcastDto, ReplyDto, etc.)
+  ↓
+DsnpAnnouncementProcessor
+  ↓ (transforms DTO → Announcement)
+Queue by Announcement Type (BROADCAST_QUEUE, REPLY_QUEUE, etc.)
+  ↓
+BatchingProcessorService (groups by queue/schemaId)
+  ↓ (Redis Lua scripts: addToBatch)
+BatchAnnouncer
+  ↓ (fetches schema from chain)
+  ↓ (converts DSNP schema → Parquet via fromDSNPSchema)
+Parquet File Creation(fromDSNPSchema arg 1)
+  ↓
+IPFS Pinning(fromDSNPSchema arg 2)
+  ↓
+On-Chain Announcement
+```
+
+## Solution 1: Schema Registry with Dynamic Discovery
+
+### Architecture Overview
+
+Replace hard-coded schema mappings with a dynamic registry that fetches schema metadata from the blockchain at startup and on-demand. This foundation enables any schema (DSNP, AT Protocol, custom) to be used without code changes.
+
+### Core Concept
+
+Instead of maintaining a static map of 20 schemas, the system will:
+1. **Query the blockchain** for schema metadata when needed
+2. **Cache aggressively** using a three-tier approach (memory → Redis → blockchain)
+3. **Auto-detect batchability** based on schema settings
+4. **Support any schema** registered on Frequency
+
+### Key Components
+
+#### 1. Schema Registry Service
+
+**Purpose**: Central service for schema metadata management and caching.
+
+**Location**: `libs/blockchain/src/schema-registry.service.ts`
+
+**Key Responsibilities**:
+- Fetch schema metadata from blockchain via existing RPC methods (`getSchema`, `getSchemaPayload`)
+- Maintain two-way caching: `schemaId ↔ SchemaDefinition` and `schemaName ↔ schemaId`
+- Determine if a schema is batchable by inspecting payload location and model type
+- Pre-load known DSNP schemas at startup for backward compatibility
+- Provide cache invalidation for testing/admin purposes
+
+**Schema Definition Interface**:
```

Comment:
Note, some of the details here will change with the new `Intents` + `Schemas` implementation